tor: add tor.proxyHook

rhendric opened the pull request:
tor.proxyHook is a setup hook intended for use with HTTP-based fetchers, enabling them to download from .onion addresses using Tor as a proxy.

Motivated by @philiptaron's suggestion in #414912.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and others READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

philiptaron reviewed (changes requested):
Requesting changes for the `tor` finalPackage thing; the rest of the suggestions are just that.

rhendric commented on pkgs/by-name/to/tor/proxy-hook.sh:
```diff
@@ -0,0 +1,19 @@
+    # Wait for Tor to start
+    read < <(<&$tor_fd- @coreutils@/bin/tee /dev/fd/2 | @gnugrep@/bin/grep -m 1 -F 'Bootstrapped 100% (done): Done')
+
+    export ALL_PROXY="socks5h://localhost$torSocket"
```
Yeah, I don't know if Tor respects `ALL_PROXY` or not, but if it does then the Tor client is already running with the old value by this point, and if it doesn't then I don't see any reasonable way to merge with the old value.

rhendric commented on pkgs/by-name/to/tor/proxy-hook.sh:
```diff
@@ -0,0 +1,19 @@
+
+    export ALL_PROXY="socks5h://localhost$torSocket"
+
+    # A Git repository may have submodules that fetch from clearnet URLs, so
+    # for better performance, use Tor only for onion addresses. (fetchgit
+    # doesn't respect ALL_PROXY, so this doesn't conflict.)
+    export FETCHGIT_HTTP_PROXIES="http://*.onion $ALL_PROXY"
```
Nah, I'll add merging logic here since I think it makes sense. Maybe multiple dark web protocols are in use in a single Git repo. Unlikely, but why not support it?

> does `http` also cover `https`?

No, but .onion addresses don't use HTTPS. The Tor protocol already includes end-to-end encryption between the Tor client and the hidden service.

rhendric committed:
tor: add tor.proxyHook

tor.proxyHook is a setup hook intended for use with HTTP-based fetchers,
enabling them to download from .onion addresses using Tor as a proxy.

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 426763 --package tor.tests`
Commit: `a3947d986ff778c6aa9f55ee9cbba33ce5019c6b`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>tor.tests.proxyHook</li>
    <li>tor.tests.tor</li>
  </ul>
</details>

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 426763`
Commit: `a3947d986ff778c6aa9f55ee9cbba33ce5019c6b`

---
### `x86_64-darwin`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>tests.fetchtorrent.http-link-rqbit</li>
    <li>tests.fetchtorrent.magnet-link-rqbit</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 61 packages built:</summary>
  <ul>
    <li>bundix</li>
    <li>cabal2nix</li>
    <li>crate2nix</li>
    <li>crystal2nix</li>
    <li>dub-to-nix</li>
    <li>gcalcli</li>
    <li>gcalcli.dist</li>
    <li>git-unroll</li>
    <li>haskellPackages.cabal2nix-unstable</li>
    <li>haskellPackages.cli-nix</li>
    <li>haskellPackages.cli-nix.doc</li>
    <li>haskellPackages.nix-thunk</li>
    <li>haskellPackages.nix-thunk.doc</li>
    <li>haskellPackages.nvfetcher</li>
    <li>haskellPackages.nvfetcher.doc</li>
    <li>haskellPackages.update-nix-fetchgit</li>
    <li>haskellPackages.update-nix-fetchgit.doc</li>
    <li>kcl</li>
    <li>lon</li>
    <li>luarocks-packages-updater</li>
    <li>luarocks-packages-updater.dist</li>
    <li>nim_lk</li>
    <li>nix-prefetch-git</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-update</li>
    <li>nix-update-source</li>
    <li>nix-update-source.dist</li>
    <li>nix-update.dist</li>
    <li>nixpkgs-manual</li>
    <li>npins</li>
    <li>nvfetcher</li>
    <li>prefetch-yarn-deps</li>
    <li>prowlarr</li>
    <li>python312Packages.nixpkgs-updaters-library</li>
    <li>python312Packages.nixpkgs-updaters-library.dist</li>
    <li>python313Packages.nixpkgs-updaters-library</li>
    <li>python313Packages.nixpkgs-updaters-library.dist</li>
    <li>radarr</li>
    <li>sonarr</li>
    <li>sus-compiler</li>
    <li>tests.fetchgit.dumb-http-signed-tag</li>
    <li>tests.fetchgit.fetchTags</li>
    <li>tests.fetchgit.leave-git</li>
    <li>tests.fetchgit.simple</li>
    <li>tests.fetchgit.sparseCheckout</li>
    <li>tests.fetchgit.sparseCheckoutNonConeMode</li>
    <li>tests.fetchgit.submodule-deep</li>
    <li>tests.fetchgit.submodule-leave-git</li>
    <li>tests.fetchgit.submodule-leave-git-deep</li>
    <li>tests.fetchgit.submodule-simple</li>
    <li>tests.fetchtorrent.http-link</li>
    <li>tests.fetchtorrent.magnet-link</li>
    <li>tests.fetchurl.header</li>
    <li>tests.haskell.cabalSdist.assumptionLocalHasDirectReference</li>
    <li>tests.haskell.cabalSdist.localHasNoDirectReference</li>
    <li>tests.testers.runCommand.bork</li>
    <li>typescript-language-server</li>
    <li>update-nix-fetchgit</li>
    <li>update-python-libraries</li>
    <li>vimPluginsUpdater</li>
    <li>yarn2nix</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-darwin`</summary>
<details>
<summary>tests.fetchtorrent.http-link-rqbit</summary>
<pre>2025-07-24T18:44:30.596682Z  INFO rqbit: starting HTTP API at http://127.0.0.1:54840
2025-07-24T18:44:30.896173Z  INFO torrent{id=0}: librqbit::session: added torrent name="The WIRED CD - Rip. Sample. Mash. Share"
2025-07-24T18:44:30.896204Z  INFO torrent{id=0}:initialize_and_start: librqbit::torrent_state::initializing: Doing initial checksum validation, this might take a while...
2025-07-24T18:44:30.896712Z  INFO torrent{id=0}:initialize_and_start: librqbit::torrent_state::initializing: Initial check results: have 0, needed 53.4Mi, total selected 53.4Mi
2025-07-24T18:44:31.598567Z  INFO rqbit: [0]: 0.00% (0 / 53.47Mi), ↓0.00 MiB/s, ↑0.00 MiB/s (0), {live: 2, queued: 24, dead: 9, known: 35}
2025-07-24T18:44:32.600022Z  INFO rqbit: [0]: 0.00% (0 / 53.47Mi), ↓0.00 MiB/s, ↑0.00 MiB/s (0), {live: 2, queued: 22, dead: 12, known: 36}
2025-07-24T18:44:34.465912Z  INFO rqbit: [0]: 0.00% (0 / 53.47Mi), ↓0.00 MiB/s, ↑0.00 MiB/s (0), {live: 3, queued: 0, dead: 34, known: 37}
2025-07-24T18:44:35.466597Z  INFO rqbit: [0]: 4.32% (2.31Mi / 53.47Mi), ↓0.05 MiB/s, ↑0.00 MiB/s (0), ETA: 1137.965s, {live: 3, queued: 12, dead: 34, known: 49}
2025-07-24T18:44:36.467897Z  INFO rqbit: [0]: 9.70% (5.18Mi / 53.47Mi), ↓0.65 MiB/s, ↑0.00 MiB/s (0), ETA: 77.339s, {live: 3, queued: 20, dead: 35, known: 58}
2025-07-24T18:44:37.468741Z  INFO rqbit: [0]: 9.70% (5.18Mi / 53.47Mi), ↓1.04 MiB/s, ↑0.00 MiB/s (0), ETA: 46.305s, {live: 4, queued: 12, dead: 44, known: 60}
2025-07-24T18:44:38.470065Z  INFO rqbit: [0]: 12.74% (6.81Mi / 53.47Mi), ↓1.12 MiB/s, ↑0.00 MiB/s (0), ETA: 42.746s, {live: 5, queued: 5, dead: 51, known: 61}
2025-07-24T18:44:39.470393Z  INFO rqbit: [0]: 21.04% (11.25Mi / 53.47Mi), ↓1.75 MiB/s, ↑0.00 MiB/s (0), ETA: 25.492s, {live: 5, queued: 8, dead: 48, known: 61}
2025-07-24T18:44:40.472037Z  INFO rqbit: [0]: 30.39% (16.25Mi / 53.47Mi), ↓2.72 MiB/s, ↑0.00 MiB/s (0), ETA: 14.578s, {live: 5, queued: 6, dead: 50, known: 61}
2025-07-24T18:44:41.472278Z  INFO rqbit: [0]: 36.47% (19.50Mi / 53.47Mi), ↓2.70 MiB/s, ↑0.00 MiB/s (0), ETA: 13.663s, {live: 5, queued: 3, dead: 53, known: 61}
2025-07-24T18:44:42.474025Z  INFO rqbit: [0]: 45.47% (24.31Mi / 53.47Mi), ↓3.31 MiB/s, ↑0.00 MiB/s (0), ETA: 9.606s, {live: 5, queued: 3, dead: 53, known: 61}
2025-07-24T18:44:43.474969Z  INFO rqbit: [0]: 49.67% (26.56Mi / 53.47Mi), ↓3.87 MiB/s, ↑0.00 MiB/s (0), ETA: 7.372s, {live: 5, queued: 10, dead: 46, known: 61}
2025-07-24T18:44:44.475241Z  INFO rqbit: [0]: 79.95% (42.75Mi / 53.47Mi), ↓4.21 MiB/s, ↑0.00 MiB/s (0), ETA: 5.628s, {live: 6, queued: 13, dead: 42, known: 61}
2025-07-24T18:44:45.476395Z  INFO rqbit: [0]: 99.65% (53.28Mi / 53.47Mi), ↓7.71 MiB/s, ↑0.00 MiB/s (0), ETA: 462ms, {live: 6, queued: 7, dead: 48, known: 61}
2025-07-24T18:44:45.945746Z  INFO torrent{id=0}:manage_peer{peer="86.22.197.178:6881"}:peer_connection: librqbit::torrent_state::live: torrent finished downloading
2025-07-24T18:44:45.945797Z  INFO rqbit: All downloads completed, exiting</pre>
</details>
<details>
<summary>tests.fetchtorrent.magnet-link-rqbit</summary>
<pre>2025-07-24T18:44:28.943468Z  INFO rqbit: increased open file limit limit=1048576
2025-07-24T18:44:28.946350Z  WARN rqbit: Error checking HTTP API at http://127.0.0.1:65391/: error sending request
2025-07-24T18:44:28.947399Z  INFO librqbit::session: Listening on 0.0.0.0:4240 for incoming peer connections
2025-07-24T18:44:28.947486Z  INFO librqbit_dht::dht: DHT listening on 0.0.0.0:59833
2025-07-24T18:44:28.947497Z  INFO librqbit_dht::dht: starting up DHT with peer id 2d7251353331302d6065af79d3f6351e870e7c66
2025-07-24T18:44:28.948439Z  INFO rqbit: starting HTTP API at http://127.0.0.1:65391
2025-07-24T18:44:29.522876Z  INFO torrent{id=0}: librqbit::session: added torrent name="The WIRED CD - Rip. Sample. Mash. Share"
2025-07-24T18:44:29.522942Z  INFO torrent{id=0}:initialize_and_start: librqbit::torrent_state::initializing: Doing initial checksum validation, this might take a while...
2025-07-24T18:44:29.950315Z  INFO rqbit: [0] initializing 0.00%
2025-07-24T18:44:30.039755Z  INFO torrent{id=0}:initialize_and_start: librqbit::torrent_state::initializing: Initial check results: have 0, needed 53.4Mi, total selected 53.4Mi
2025-07-24T18:44:31.040027Z  INFO rqbit: [0]: 19.40% (10.37Mi / 53.47Mi), ↓0.00 MiB/s, ↑0.00 MiB/s (0), {live: 5, queued: 19, dead: 8, known: 32}
2025-07-24T18:44:32.041738Z  INFO rqbit: [0]: 53.30% (28.50Mi / 53.47Mi), ↓10.44 MiB/s, ↑0.00 MiB/s (0), ETA: 4.128s, {live: 5, queued: 22, dead: 9, known: 36}
2025-07-24T18:44:33.042176Z  INFO rqbit: [0]: 90.42% (48.34Mi / 53.47Mi), ↓14.28 MiB/s, ↑0.00 MiB/s (0), ETA: 1.748s, {live: 5, queued: 3, dead: 28, known: 36}
2025-07-24T18:44:33.863134Z  INFO torrent{id=0}:manage_peer{peer="98.15.176.57:58899"}:peer_connection: librqbit::torrent_state::live: torrent finished downloading
2025-07-24T18:44:33.863819Z  INFO rqbit: All downloads completed, exiting</pre>
</details>
</details>

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 426763 --package tor.tests`
Commit: `a3947d986ff778c6aa9f55ee9cbba33ce5019c6b`

---
### `x86_64-darwin`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>tor.tests.proxyHook</li>
    <li>tor.tests.tor</li>
  </ul>
</details>

rhendric commented:
Thanks for the guidance, @philiptaron!
